### ISSUE TITLE
fsspec Source (the local file one, at least) does not close files

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/uproot5#1292]
I saw this in @giedrius2020's talk.

```python
>>> import uproot
>>> import skhep_testdata
>>> with uproot.open(skhep_testdata.data_path("uproot-Zmumu.root")) as file:
...     tree = file["events"]
... 
>>> tree["px1"].array()
<Array [-41.2, 35.1, 35.1, 34.1, ..., 32.4, 32.4, 32.5] type='2304 * float64'>
>>> tree.file.closed
False
```

The `with` statement was supposed to close the file, making it impossible to later read the arrays from it. (Without this, the `with` statement is useless: its purpose is to prevent file-handle leaks, and nothing here is closing the file-handle.)

This is where the "`file`" object (a ReadOnlyDirectory) has an `__exit__` method that gets called at the end of the `with` statement:

https://github.com/scikit-hep/uproot5/blob/dc19ce911caf45b3239d47aa61e1732f79417b29/src/uproot/reading.py#L1515-L1516

and here is where that gets propagated to an FSSpecSource:

https://github.com/scikit-hep/uproot5/blob/dc19ce911caf45b3239d47aa61e1732f79417b29/src/uproot/source/fsspec.py#L81-L83

and here's the pre-fsspec Source (MemmapSource):

https://github.com/scikit-hep/uproot5/blob/dc19ce911caf45b3239d47aa61e1732f79417b29/src/uproot/source/file.py#L216-L223

Pre-fsspec file-handles got closed:

```python
>>> import uproot
>>> import skhep_testdata
>>> with uproot.open(skhep_testdata.data_path("uproot-Zmumu.root"), handler=uproot.MemmapSource) as file:
...     tree = file["events"]
... 
>>> tree["px1"].array()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/irishep/uproot5/src/uproot/behaviors/TBranch.py", line 1825, in array
    _ranges_or_baskets_to_arrays(
  File "/home/jpivarski/irishep/uproot5/src/uproot/behaviors/TBranch.py", line 3023, in _ranges_or_baskets_to_arrays
    hasbranches._file.source.chunks(ranges, notifications=notifications)
  File "/home/jpivarski/irishep/uproot5/src/uproot/source/file.py", line 163, in chunks
    raise OSError(f"memmap is closed for file {self._file_path}")
OSError: memmap is closed for file /home/jpivarski/.local/skhepdata/uproot-Zmumu.root
>>> tree.file.closed
True
```

Now I wonder why pytest is not complaining about the leaking file-handles. I know that it checks for unclosed files because I've seen that error. Here it is, reproduced on the terminal, and our test suite never raised an error. I don't understand that.

Socially, this could be a problem because the fsspec-based file-handles have been available for almost a year (Uproot 5) and users might have gotten used to opening the file in a `with` statement (uselessly) and later, outside the `with`, initiating more file-reading with TTree, RNTuple, and TDirectory objects. After fixing this, code that relied on this error will break. It will need to be a new minor version number, 5.4.0, at least.

**Comment by ariostas** [comment]
I was curious about this, so here's what I found out.

The `fsspec` implementation is set up to always return `False` for the closed property.
https://github.com/scikit-hep/uproot5/blob/53f917c82daac2f07e9a600226a1c358783a1653/src/uproot/source/fsspec.py#L193-L201

But if you check the underlying file `tree.file.source._file.f.closed`, then you find that it was closed. It is weird that it is still being able to read the file since it's supposed to be closed.

I feel like the comment above is explains the existence of the property, but not why it always returns `False`, so I'm guessing it was just a silly mistake. I can submit a PR to simply replace it to `return self._file.f.closed` if you agree.

**Comment by jpivarski** [comment]
What we can do is make final releases of Awkward and Uproot with Python 3.8 support, and then immediately afterward make a _minor_ release without Python 3.8 support, so that the change in Awkward version number only changes that level of support.

The question then becomes, which PRs do you want to be sure are included in the last with-Python-3.8 release?

**Comment by ariostas** [comment]
>The question then becomes, which PRs do you want to be sure are included in the last with-Python-3.8 release?

I think it would make sense to at least wait for the fix/perf PRs that are ready (or almost ready). There's no reason to rush to drop Python 3.8, but I did notice that things started to break for 3.8 (see e.g https://github.com/scikit-hep/scikit-hep-testdata/pull/162).
